metomi/rose/loc_handlers/git.py

```diff
@@ -0,0 +1,191 @@
+# Copyright (C) British Crown (Met Office) & Contributors.
+# This file is part of Rose, a framework for meteorological suites.
+#
+# Rose is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Rose is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Rose. If not, see <http://www.gnu.org/licenses/>.
+# -----------------------------------------------------------------------------
+"""A handler of Git locations."""
+
+import os
+import re
+import tempfile
+from urllib.parse import urlparse
+
+
+REC_COMMIT_HASH = re.compile(r"^[0-9a-f]+$")
+
+
+class GitLocHandler:
+    """Handler of Git locations."""
+
+    GIT = "git"
+    SCHEMES = [GIT]
+    WEB_SCHEMES = ["https"]
+    URI_SEPARATOR = "::"
+
+    def __init__(self, manager):
+        self.manager = manager
+        ret_code, versiontext, stderr = self.manager.popen.run(
+            "git", "version")
+        if ret_code:
+            self.git_version = None
+        else:
+            version_nums = []
+            for num_string in versiontext.split()[-1].split("."):
+                try:
+                    version_nums.append(int(num_string))
+                except ValueError:
+                    break
+            self.git_version = tuple(version_nums)
+
+    def can_pull(self, loc):
+        if self.git_version is None:
+            return False
+        scheme = urlparse(loc.name).scheme
+        if scheme in self.SCHEMES:
+            return True
+        if self.URI_SEPARATOR not in loc.name:
+            return False
+        remote = self._parse_name(loc)[0]
+        return (
+            scheme in self.WEB_SCHEMES
+            and not os.path.exists(loc.name)  # same as svn...
+            and not self.manager.popen.run(
+                "git", "ls-remote", "--exit-code", remote)[0]
+            # https://superuser.com/questions/227509/git-ping-check-if-remote-repository-exists
+        )
+
+    def parse(self, loc, conf_tree):
```

Comment:
The parse and pull operations overlap heavily duplicating some of the same operations which exacerbates efficiency (and potential rate limiting) issues.

Ideas:
* Re-use the same temporary directory and tidy it up later.
* Skip parse checks and make parse-time errors run-time errors.

Reply:
We have now gone for option (2) and have removed the parse-time checks by using a trailing slash to denote the difference between a file and a directory. This gets checked at run-time.